docs/Migration Guide.md

```diff
@@ -3,6 +3,146 @@
+index.
+If you have WASM frames, you'll need to handle those separately.
+
+See [bug 1847469](https://bugzilla.mozilla.org/show_bug.cgi?id=1847469).
+
+### New API for ArrayBuffer buffers ###
+
+ArrayBuffer APIs that take data buffers from the embedding, such as
+`JS::NewExternalArrayBuffer()` and `JS::NewArrayBufferWithContents()`,
+now take a `mozilla::UniquePtr&&` argument for the data buffer.
+
+**Recommendation:** This is a fairly straightforward migration, in which
+you can just put your data buffer into a UniquePtr and then
+`std::move()` it into the ArrayBuffer.
+If you previously passed a free callback, use that callback as the
+deleter argument to the UniquePtr.
+Examples:
+
+```c++
+// old
+JS::RootedObject array_buffer{cx, JS::NewExternalArrayBuffer(
+  cx, len, my_struct->buffer, my_buffer_free_func, my_struct)};
+
+// new
+mozilla::UniquePtr<void, JS::BufferContentsDeleter> contents{
+  my_struct->buffer, {my_buffer_free_func, my_struct}};
+JS::RootedObject array_buffer{cx, JS::NewExternalArrayBuffer(
+  cx, len, std::move(contents))};
+```
+
+### `JS::GCPolicy::needsSweep()` ###
+
+`JS::GCPolicy<T>` has a new method, `needsSweep()`.
+It is like a const version of `JS::GCPolicy::traceWeak()`, but the sense
+of the boolean return value is reversed.
+(Returns true if the weak pointer is dead, and false if it is live.)
+The argument is const, so the method will not be called if the pointer
+needs to move.
+Implementing the method is optional.
+
+**Recommendation:** If you have specialized `JS::GCPolicy` for any of
+your classes, consider whether you need to implement this method.
+If your class has a weak pointer, you probably do.
+
+Also, note that `JS::GCPolicy<JS::Heap<T>>` does not implement
+`needsSweep()`.
+If you were storing a `JS::Heap` inside `JS::WeakCache<T>`, you will
+need to use a different data structure.
+An easy drop-in replacement for `JS::Heap` suitable for this purpose
+would be something like this:
+
+```c++
+template <typename T>
+class WeakPtr : public JS::Heap<T> {
+ public:
+  using JS::Heap<T>::Heap;
+  using JS::Heap<T>::operator=;
+};
+
+namespace JS {
+
+template <typename T>
+struct GCPolicy<WeakPtr<T>> {
+  static void trace(JSTracer* trc, WeakPtr<T>* thingp, const char* name) {
+    return JS::TraceEdge(trc, thingp, name);
+  }
+
+  static bool traceWeak(JSTracer* trc, WeakPtr<T>* thingp) {
+    return js::gc::TraceWeakEdge(trc, thingp);
+  }
+
+  static bool needsSweep(JSTracer* trc, const WeakPtr<T>* thingp) {
+    WeakPtr<T> thing{*thingp};
+    return !js::gc::TraceWeakEdge(trc, &thing);
+  }
+};
+
+}  // namespace JS
+```
+
+### Various API changes ###
+
+This is a non-exhaustive list of minor API changes and renames.
+
+- `JS_GetFunctionId()` is renamed to `JS_GetMaybePartialFunctionId()`.
+  There is now a different API named `JS_GetFunctionId()` that requires
+  a JSContext.
+- Same for `JS_GetFunctionDisplayId()` and
+  `JS_GetMaybePartialFunctionDisplayId()`.
```

Comment:
`JS_GetMaybePartialFunctionId` is different function than the old `JS_GetFunctionId`.
The old `JS_GetFunctionId`'s behavior is preserved in the new `JS_GetFunctionId`, but with different signature.

`JS_GetMaybePartialFunctionId` can return a function name that lacks the `"get "` or `"set "` prefix for accessors.

This is for performance reason, where the accessor function name is rarely accessed, and computing the function name is almost always unnecessary.

This happens if the accessor is defined via `JS_DefinePropertyById`, `JS_DefinePropertyById`, or `JS_DefineProperties`.
(to be clear, this happens only for accessors)

Functions with "MaybePartial" can be used only for debugging purpose where the name doesn't have to be exact.

Reply:
Thanks for the explanation. I've split this out into its own section now, since it is a bit more complicated than I thought.